mllib/src/main/scala/org/apache/spark/ml/param/shared/sharedParams.scala

```diff
@@ -274,6 +274,25 @@ trait HasOutputCols extends Params {
   final def getOutputCols: Array[String] = $(outputCols)
 }
 
+/**
+ * Trait for shared param numFeatures (default: 1 << 18). This trait may be changed or
+ * removed between minor versions.
+ */
+@DeveloperApi
+trait HasNumFeatures extends Params {
+
+  /**
+   * Param for number of features..
+   * @group param
+   */
+  final val numFeatures: IntParam = new IntParam(this, "numFeatures", "number of features.", ParamValidators.gt(0))
+
+  setDefault(numFeatures, 1 << 18)
+
+  /** @group getParam */
+  final def getNumFeatures: Int = $(numFeatures)
```

Comment:
Hm, it was introduced in the subclasses in different versions. We could say `@Since("2.3.0")` at least?

Reply:
Current `SharedParamsCodeGen` do not support since annotation. What about adding this feature in `SharedParamsCodeGen` in the future?